### ISSUE TITLE
“Type instantiation is excessively deep and possibly infinite” but only in a large codebase

### ISSUE THREAD

**Issue by karol-majewski** [microsoft/TypeScript#34933]
**TypeScript Version:**  3.7.2, 3.8.0-dev.20191102 (worked in 3.6)

**Search Terms:**

- Type instantiation is excessively deep and possibly infinite.ts(2589)
- Mapped types
- Generics
- Conditional types

**Code**

Note: **this issue manifests itself only in our codebase**. When you run the same code in TypeScript Playground, it seems to be working fine.

The snippet is hardly minimal, but I reduced it as much as I could. I recorded [a video](https://streamable.com/pjkrk) where exactly the same code yields an error different than the one in TypeScript Playground. I tried with two versions of TypeScript: `3.7.2` and `3.8.0-dev.20191102`. It worked correctly with `3.6`.

Since @sheetalkamat and @DanielRosenwasser have access to our repository, you're welcome to have a look at [this PR](https://github.com/unsplash/unsplash-web/pull/4243). Copy-paste the code below anywhere in the project to see the error.

The versions of types used:

- `@types/history@4.7.3`
- `@types/react@16.9.11`
- `@types/react-router-dom@5.1.0`
- `@types/recompose@0.30.7`

**Note**: Interestingly enough, if you change:

```diff
- declare const Button: React.FunctionComponent<Omit<Props, never>>;
+ declare const Button: React.FunctionComponent<Props>;
```

it works again despite the fact `Omit<Props, never>` should be the same as just `Props`.

<details>

<summary>Source code</summary>

```ts
import { History } from 'history'; // "4.7.3"
import * as React from 'react'; // "16.9.11"
import { LinkProps, RouteComponentProps, withRouter } from 'react-router-dom'; // "5.1.0"
import { getDisplayName } from 'recompose'; // "0.30.7"

declare function isDefined<T>(candidate: T | null | undefined): candidate is T;
declare function isString(value?: any): value is string;

type ObjectOmit<T extends K, K> = Omit<T, keyof K>;

type OnClick = NonNullable<React.ComponentProps<'button'>['onClick']>;

type OnClickProp = {
  /** If there is a custom click handler, we must preserve it. */
  onClick?: OnClick;
};

type ProvidedProps = OnClickProp;

type InputProps = OnClickProp & {
  /** Note: we want this helper to work with all sorts of modals, not just those backed by query
   * parameters (e.g. `/photo/:id/info`), which is why this must accept a full location instead of a
   * `Modal` type.
   * */
  to: Exclude<LinkProps['to'], Function>;
};

const buildClickHandler = ({
  to,
  onClick,
  history,
}: InputProps & {
  history: History;
}): OnClick => {
  const navigate = () => {
    // https://github.com/Microsoft/TypeScript/issues/14107
    isString(to) ? history.push(to) : history.push(to);
  };

  return event => {
    [onClick, navigate].filter(isDefined).forEach(callback => callback(event));
  };
};

/** See the test for an example of usage. */
export const enhance = <ComposedProps extends ProvidedProps>(
  ComposedComponent: React.ComponentType<ComposedProps>,
) => {
  type PassThroughComposedProps = ObjectOmit<ComposedProps, ProvidedProps>;
  type OwnProps = InputProps & RouteComponentProps<never> & PassThroughComposedProps;
  type Props = OwnProps;

  const displayName = `CreateModalLink(${getDisplayName(ComposedComponent)})`;

  const ModalLink: React.FunctionComponent<Props> = ({
    to,
    onClick,

    history,
    // We specify these just to omit them from rest props below
    location,
    match,
    staticContext,

    ...passThroughComposedProps
  }) => {
    const clickHandler = buildClickHandler({ to, onClick, history });

    const composedProps: ComposedProps = {
      // Note: this is technically unsafe, since the composed component may have props
      // with names matching the ones we're omitting.
      // https://github.com/microsoft/TypeScript/issues/28884#issuecomment-503540848
      ...((passThroughComposedProps as unknown) as PassThroughComposedProps),
      onClick: clickHandler,
    } as ComposedProps;

    return <ComposedComponent {...composedProps} />;
  };

  ModalLink.displayName = displayName;

  return withRouter(ModalLink);
};

type Props = React.ComponentPropsWithoutRef<'button'> &
  Required<Pick<React.ComponentPropsWithoutRef<'button'>, 'type'>>;

/**
 * This one errors.
 */
declare const Button: React.FunctionComponent<Omit<Props, never>>;

/**
 * This one works.
 */
// declare const Button: React.FunctionComponent<Props>;

const EnhancedButton = enhance(Button);

/**
 * Type instantiation is excessively deep and possibly infinite.ts(2589).
 */
() => <EnhancedButton></EnhancedButton>;


```

</details>

**Expected behavior:**

I should get a proper error about missing properties (not the one about type instantiation):

```
Type '{}' is missing the following properties from type 'Readonly<Pick<OwnProps, "form" | "style" | "title" | "onClick" | "to" | "key" | "autoFocus" | "disabled" | "formAction" | "formEncType" | "formMethod" | "formNoValidate" | "formTarget" | ... 252 more ... | "onTransitionEndCapture">>': to, type(2739)
```

**Actual behavior:**

I'm getting this:

```
Type instantiation is excessively deep and possibly infinite.ts(2589).
```

**Playground Link:** 

[Playground Link](https://www.typescriptlang.org/play/?jsx=2#code/JYWwDg9gTgLgBAbzgCWAZxtAnnAvnAMyghDgHIALdTKLMgbgChRJY4AqOAQzTgCUAplwDG8IiXJQhohs3DR4SADLAAdgGsACsTBoANPwgBXGAIDCJSKoGqY2iLoMB3YDAp9jpqHkLFSZKREYAFpiEwEoYIATElkWBUQ4AHMBGAARdDAAGy4sADkuEAEfcX8pYUsINAFZRiiBYRypQiNVUWAIVTh0NIECNQEogB4AFQA+AAphLlUo4CiuUwAuOBG4AB84VSMsrI24Vvr+6yiAShXp2fnF4vRVpnrGrmaCVvbO7rQAZRgoNSSJgA3LhZIwCAD8KxmWHOcGBoNuvAwf1USSYjBgWDAxQA8gAjABWDRgOJArlGcAEAA9TLNeABpAz0sZwAC8cFJ5JGBnUAiwEAIcGZ6Mx2I5qjMWWAwnUbLgeU6eR2OTxWQEQ0EQQAdBZ5NZbPZdEMyHiTJhVGQxgBtMidSXS9RkAC6YxFWNxEqlMsNcoQjDgcAA9OxOABJQVuCKI7hwYRGDASRoOuAUGZRNVQZzFEDx+BgKTVKCA24wLUcQP+uB2r3qSHi+0ypi4N1i+yA+aDQ28dk4z0Ow0t4qh1RgExduW9htaHRwABkiErwc4CuWcCcxScM3gbjuFAEWWx3kwa+gspcbm4uzgaAUvAFcBAEAWWX0Wwg8AJubgbiqxTxIl5KI4DxHAAEcwVoSsODgMBnkKVIIl4CYBC1JIywAA0DMAKHfCBAyWeZAzUAgIHQ05nCoYQKE+NcKBwHdeBzDBuGEYQBDAeAuBaK8sggaYYA6Lo1AwIQgPvLgoM4dCAFknxBdDv3dLVJPLStMBWABRKlGiMeohhUDQuxtTBnQMAAxN4BM6V1GGbRhGAqVRmNNYAsiiKdkDTDM5QmP0A0wPRK2rB1AoDKgE1oQLcBWYdRzsHReHnPyU2obAVlQCKsCbWFJxrNkWWSxzmNULh2ySG4fNOfKFwDANA0DFMYBgXQlnqpJXAoIw8S1CoQEDaTpWIG8CBgQMRndL5hD+DiiLQNAwTQQMAEYABYloABgAdigz4fhRAFMCq8EUsyrVRzQCgJkOuAVnCmgsDO+NLsOpgAzsyspBgIwoC6ARi1sarkoDK1gplAwSrKm4nS1fosi8CYej6AYzhh6ANJES7pl2f8ZWqrGshx9RkP+mBTlOV68CbdElzgL4BGKSNvwEZiSO8GZKSpQpsmKe94y4FIy3YCtqVYeAivgGxUzaYp2SGXVIGqKJx2pWkol4NsOyVhLJkreXf3cyp9RgFZNVEHVDZsGBxuxOXKkVrsxkCqrWQKtT3TgTQeDQEYKDCJIKD1+2EonQliU5GBbfkIOHFfDX6i1mObP892cScVRx3ZWKx2D+cPHCPWja7IZrGLKAWXnT25p9v2A7tzsEop0Vigzjk067dEA3FuA5jQbJcgKIo5XQsxAlMWTnwMomABIEBSdJMhyfJ4ImQPBgLy3TlwU50I72NOmY8eQUnk3pFLCy2isiULdsIYHZ8oHvwgULatB9RAp2u7sGfuqGoAdWKXuDRgAEHonuaocBPzMWPCQVw349ykFKHAAseZg54n3BAJwO1eL8UEt-B8iwqJ4IwIsaUFhbAq3frVOAWoaGwSrr7Yw-tV4J10JWLegMdpdyTDKTysxvLshcm5DyXkIi+UfgYV+BhP60DwOTeyVCuF1xYWgFYzCW4Px-vKd8AgVgMRoqYKiqhpQgiyDgVoaAuAEAEAYNAag2JwOKL1BWgw956ktvgnAqZiwwQSjtTR55qIlSKIxAhVBUQOKrNYXg64Ag8zJE1f4ykqGaIoE1FqbUOpdR6iQQMZIppVAFKNa2AhJrTVGugeazNAwACYAAc9SVoAGIKlgl6kUWwwQACs60ADMnSVrrVqStWpfjqE0ImBMOh3sGFGCYUo8cPADgaFUBg1QVVFmV2mTXNRCVyKjNfhcGsvD0wRDwfgRZOyY67wDJ9b6XRI7OINm4gGCAaHZKjvXGO+BAxJ0pvIuAh8siTy1D3PuS9B7slBYvAeAhd63J+muDqed4aAsnnI96TcPbB3ZKbUs68DQJV-h1TwggCDGlNE1Tolo5yVkEOBYAUhhiaAdBqU+5tnnxRjkSn8JhSXkrNFSx25Am6WhsowJc-pOA+zuJ0YoERiBQDQEkoWdQGhNEcfveAAAhAVqgT7anPu8K+HKhjh1vglcGf0IhjDFRK6C0q7zWBPFAdQSrJUVnqt3NVzwNVOW1bq-VZtDWX3xRHB26Iu4aVUFLNiUQdWUq6OySWMw2ITHjeaOR4qQyStWO7YSMAtzABIR8O41I2JzWAMWUxXr2LcFmDBKotjVQ4GImoVwKEYBoAmNUzptSACcpxlUVgmM7FkQwo0xsGOm6yQxAwTpTVO3VYqgA)

**Related Issues:** 

- #32735
- #34850

**Comment by AnyhowStep** [comment]
>  | ... 252 more ... | 

That's a lot of properties ._.

**Comment by karol-majewski** [comment]
@AnyhowStep That's correct. The component in question mixes-in the props defined by `JSX.IntrinsicElements['button']`. There is a lot going on, and it is complicated, but it's a regression nonetheless because it worked in TypeScript 3.6.3.

**Comment by Stianhn** [comment]
We have the same issue in one of our larger projects. Works fine with 3.6.4 but not with 3.7.2

**Comment by rista404** [comment]
Same issue, using `keyof JSX.IntrinsicElements` type.

**Comment by peterkelly** [comment]
in `src/compiler/checker.ts` on line 13208 (in v3.7.2) there is the following code:

```
function instantiateType(type: Type | undefined, mapper: TypeMapper | undefined): Type | undefined {
    if (!type || !mapper || mapper === identityMapper) {
        return type;
    }
    if (instantiationDepth === 50 || instantiationCount >= 5000000) {
        // We have reached 50 recursive type instantiations and there is a very high likelyhood we're dealing
        // with a combination of infinite generic types that perpetually generate new type identities. We stop
        // the recursion here by yielding the error type.
        error(currentNode, Diagnostics.Type_instantiation_is_excessively_deep_and_possibly_infinite);
        return errorType;
    }
    instantiationCount++;
    instantiationDepth++;
    const result = instantiateTypeWorker(type, mapper);
    instantiationDepth--;
    return result;
}
```

So the value is hard-coded. Perhaps a configuration option to change this might be warranted. As a workaround for now, you could try upping this limit and rebuilding the compiler.

**Comment by AnyhowStep** [comment]
There have been proposals made to increase the limit. They've been rejected because it's seen as an "implementation detail" that should not be exposed.

**Comment by karol-majewski** [comment]
@Stianhn @rista404 Have you managed to reproduce this issue using a more reduced example? If what we're dealing with is a valid use case, then understanding it could help us come up with a heuristic for when type instantiation limit should not be applied.

I agree that messing with the internals leads nowhere — after all, you need to draw the boundary somewhere. 

As far as I know, this limit is not something new. It existed before, yet somehow this code stopped working. I wonder what changed that directly influenced our use cases.

**Comment by AnyhowStep** [comment]
git bisect? =x

**Comment by DanielRosenwasser** [comment]
Can you grant @weswigham permission to the repository as well?

**Comment by karol-majewski** [comment]
@DanielRosenwasser Yes! Invite sent @weswigham.

**Comment by dgieselaar** [comment]
Another example in https://github.com/elastic/kibana/pull/47188/checks?check_run_id=303150613 where @timroes is trying to upgrade TS to 3.7.2:

```
x-pack/legacy/plugins/apm/server/lib/metrics/by_agent/java/heap_memory/index.ts:58:10 - error TS2589: Type instantiation is excessively deep and possibly infinite.

  58   return fetchAndTransformMetrics({
              ~~~~~~~~~~~~~~~~~~~~~~~~~~
  59     setup,
    ~~~~~~~~~~
... 
  70     additionalFilters: [{ term: { [SERVICE_AGENT_NAME]: 'java' } }]
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  71   });
```

Related pull request:
https://github.com/elastic/kibana/pull/47188

Related file:
x-pack/legacy/plugins/apm/server/lib/metrics/fetch_and_transform_metrics.ts

**Comment by AnyhowStep** [comment]
Worth noting that if you're upgrading from TS 3.5, something about how `constraintDepth` works was changed that can cause the max depth error, for code that was previously working fine (and nesting 100+ layers deep).

https://github.com/microsoft/TypeScript/issues/33541

**Comment by fabb** [comment]
We have a similar issue with this code (roughly):

```
export type IconProps = TestProps & SizeProps & SpaceProps & DisplayProps & ColorProps<any>

// these are more than 450 props
export type ReactSvgProps = Omit<SVGProps<SVGSVGElement>, 'height' | 'width'>

export const createSvgIcon = (Component: FunctionComponent<ReactSvgProps>) => styled(Component)<IconProps>({display: 'inline-block', flex: '0 0 auto' }, compose(size, space, display, color))

// here we get the error
const MyIcon = createElement(Icon, { color: 'blue', marginRight: 'xs', marginLeft: '-xs'  })
```

**Comment by macmillan78** [comment]
I have a similar issue with grommet-icons v4.4.0:

If I use the trash icon directly, all fine. If I wrap it with styled from styled component, I get the same message. No problem until typescript 3.5.2.

No Problem:

```
import { FormTrash } from 'grommet-icons';
export const AdProgramView = (props: PropType) => {
    return (
        <Container>
            <h3>
               {props.headline}
                <FormTrash onClick={deleteIntentCallback} />
            </h3>
        </Container>
    )
}
```

Problem with 3.7.2:

```
import { FormTrash } from 'grommet-icons';
const FormTrashStyled = styled(FormTrash)`
  vertical-align: bottom;
  cursor: pointer;
`
export const AdProgramView = (props: PropType) => {
    return (
        <Container>
            <h3>
               {props.headline}
                <FormTrashStyled onClick={deleteIntentCallback} />
            </h3>
        </Container>
    )
}
```

**Comment by j-o-d-o** [comment]
Same issue with styled components extending rmwc components:
```javascript
import * as React from 'react'
import styled from 'styled-components'
import { Button } from '@rmwc/button'

const ButtonStyled = styled(Button)`
`

export function ConnectionSetting() {
  return <React.Fragment>
    <ButtonS />
  </React.Fragment>
}
```

**Comment by sanketjo96** [comment]
So what is the fix here. Nothing yet ? and we should not update to latest ts ? or refactor application code within limits ?

**Comment by mkantz84** [comment]
I've noticed that strict mode fixes it.
I just added `"strict": true` to `tsconfig.json` and it solved

**Comment by ds82** [comment]
> I just added `"strict": true` to `tsconfig.json` and it solved

That does not solve it for me :/

**Comment by qtiki** [comment]
I ran into this with [xstate](https://github.com/davidkpiano/xstate). The types work with 3.7.5 but upgrading to 3.8.2 gives `error TS2589: Type instantiation is excessively deep and possibly infinite.`

I made a quick gist to reproduce the issue, hope it helps: https://gist.github.com/qtiki/5f97233f609e516b5439e926514e85d9

Download the gist and run `yarn` and `yarn tsc` to get the error (or some `npm` commands if you prefer).

**Comment by vshlos** [comment]
Here is a small snippet that causes the issue. It is recursive but not using React.

```

export declare type NoMethods<T> = Pick<T, Exclude<{
    [P in keyof T]: T[P] extends Function ? never : P;
}[keyof T], undefined>>;

export type PC<T, TChild=T> = {
    [K in keyof NoMethods<TChild>]-?: PC<T, Exclude<TChild[K], undefined>>
} & {get(): string}


function test<T>(values: PC<T, unknown>[]): string[] {
    return values.map(curr => curr.get()); //ERROR HERE
}
```
The strange thing is if I remove the map and change it to a forEach, this works fine. As in:
```
function test<T>(values: PC<T, unknown>[]): string[] {
    const results: string[] = [];
    values.forEach(curr => {
        results.push(curr.get());
    });
    return results;
}
```
has no errors.

**Comment by Leandro-Albano** [comment]
I have one small reproduction snippet too:
```
export type KnownKeys<T, TType = any> = {
  [K in keyof T]: string extends K ? never : number extends K ? never : K;
} extends { [_ in keyof T]: infer U }
  ? U extends keyof T
    ? T[U] extends TType
      ? U extends string
        ? U
        : never
      : never
    : never
  : never;

export interface QueryBuilder<T, TResult = T> {
  select<K extends KnownKeys<TResult>>(...keys: K[]): QueryBuilder<TResult, { [k in K]: TResult[k] }>;
}

export class QueryableInclude<T, TResult = T> implements QueryBuilder<T, TResult> {

  select<K extends KnownKeys<TResult>>(...keys: K[]): QueryableInclude<T, { [k in K]: TResult[k] }> {
    throw new Error("Method not implemented.");
  }

}

```

It seams to work fine on **3.3.3**, but fail in all subsequent versions.
On the concrete type (`QueryableInclude`) it I return the interface instead of the concrete, it works fine. If is make `KnownKeys` less complex it works fine too.

**Comment by moltar** [comment]
I'm getting it on the following snippet:

```ts
export class BaseConfig<T extends typeof BaseConfig> {
  public constructor(private readonly environment: Pick<InstanceType<T>, keyof T>) {
    // ...
  }
}
```

I was attempting to get the public props as constructor arguments type.

**Comment by chicken-suop** [comment]
https://stackoverflow.com/questions/58311182/type-instantiation-is-excessively-deep-and-possibly-infinite-with-emotion-styled

**Comment by mateja176** [comment]
Ran into the problem implementing a recursive mapper type:

```ts
export interface Route {
  path: string;
  label: string;
  icon: React.ReactElement;
}

export type Routes = Record<string, Route>;

export interface NestedRoute extends Route {
  children: NestedRoutes;
}

export type NestedRoutes = Record<string, NestedRoute>;

export type GetNestedRouteKey<R extends NestedRoutes> = {
  [key in keyof R]:
    | key
    | keyof R[key]['children']
    | GetNestedRouteKey<R[key]['children']>;
}[keyof R];

// * Example
type A = {
  ideas: {
    path: 'ideas';
    label: 'Discover';
    icon: ReactElement;
    children: {
      create: {
        path: 'create';
        label: 'Create';
        icon: ReactElement;
        children: {};
      };
      my: {
        path: 'my';
        label: 'My Ideas';
        icon: ReactElement;
        children: {
          published: {
            path: 'published';
            label: 'Published';
            icon: ReactElement;
            children: {};
          };
        };
      };
    };
  };
};
type B = GetNestedRouteKey<A>; // "ideas" | "create" | "my" | "published"

export type FlattenRoutes<R extends NestedRoutes> = Record<GetNestedRouteKey<R>, Route>; // Type instantiation is excessively deep and possibly infinite.

```

**Comment by yonilerner** [comment]
I get this issue right now but only in watch mode and only after file changes

**Comment by C0ZEN** [comment]
Got the issue with the latest `discord.js` release (12.3.0) and `ts-auto-mock` (2.3.4).
More details on the [ts-auto-mock issue](https://github.com/Typescript-TDD/ts-auto-mock/issues/471).

**Comment by callionica** [comment]
I have isolated an example of hitting this error when trying to provide types for a function that deals with async iterables. Error is only hit when the function implementation is available and replacing the implementation with a declaration removes the problem even though the types have to be instantiated all the same.

The recursive type looks like this.

type ZipResult<T extends unknown[], Result extends unknown[] = []> = {
    0: ZipResult<Tail<T>, Prepend<(InferIteratorValue<Head<T>> | undefined), Result>>,
    1: Reverse<Result>
}[Length<T> extends 0 ? 1 : 0];

The input array that instantiates this type is two elements long, so, perhaps naively,  I would expect the recursion depth to be 6 (3 for recursion into ZipResult itself and 3 for recursion into Reverse). Nowhere near the limit of 50 mentioned earlier in the thread.

Code is available in the playground:

https://www.typescriptlang.org/play?target=6&jsx=0&ts=4.0.2#code/PQKhCgAIUgVALAlgZ0gYwPYBMCmlcC2GAdsgC4BOAhmTqlZAESwCeADnoqWVcWYjUQlIKSDgAeaOskQA3HABsW+HDjaReWSGwzIZAIyUjiAMy6JajMRQoYKxyLIAsAOgAMLgExQYPuPDwAcxxiHGpaLQApAGV0bDwKAFdSOJscNDIlFz8EPDJ2OkhkeAwAd2IHMgDIMgwMBX42VAU6gGtIAHkAaWzofzwSIzZbQxwCEVQqvDDbCl7fPoBJMkgCKmV9BMUaHC1ayESZYkCNCqpkFmI0SGDQ8LtIEwepkQI2BTGQnn4SAH5-vwAJTGGFkXBOL0Qbw+BC+gmEmkgFDUCioaHBIhWpQs8FOYgAjok5FQPnwVGhUeEhBUABQSKRsFZPewvEAASjECihXB2k2qMzsPh8wHA4HyHEgi1MYWWYRodgAaiTEjgADywAB8kAAvFBIPrYGJxLRiFhULLqKNVVwTGFIAq3Frfva3JAAFx6g1Gk1myW0e4Ua3S+wKgCMTvtofdnrg3pCvoAghcrhaqFabXaFZ4I1no-qvRIfagk5c0KnaoGMyGAMw56t5-OheQUADcovFeAAMiFAlV1XHTahkq1iGViABtAC6Wu1cHHAHJSb34PPJ22xQVIAAJHBULD9wvxofEEdjqcz2OHweQcfD0flAA0kBcL7vZ8nk8gztg47cn7dkBNmE64dnAVCIAoB7GkeBwnveE7TjqMbAMAKijgAtFyfDoYggSjsigEYOhEjvIg6JkOhvAsDGNI0i+LhkABsActqWpvuUHJXr6tGMacLBPvRPAQQBVYuixbFwWOHK-DGDoNvqQGtu2m4AArIhwpqqgAok+hpccep7lOeSH5pAtEBHuAE6c+L5UBQgTIEx4mwYZxCcdB160fRdkOSJwZiTqEmudJLrunAIGbsCzbIGqekeb67EIU+HQDglklGZ+s7GbOADeMZuABUVhDF6rgZBmpPmpajxqqO57uqGrJRqGoxqGAEdOAAC+47dscfaaqlqCus6UYAX+EUSgAWogbDAsgiQNFBRYufBU5PnNC0rPpK3vjqN6Ibl+UAdNs10JtpUQQ1lXqTVNJSraFDloqypqnV+6alqAA+sG4GYoRYGy61nQ0zUPq1hU4NFaobSDXU9T2-Vattw2QKNkDjaKKGQFpgKAh0gJMZuXDkLw-DwhUoj0tIciKMouBqKcWg6HoiCGMoNrmLQ4BY7gGFYRRuH4XgGEkVy5GUcQLDoQAIos0QJgAQp2WnS+A5ylo8yQZNSMAAF4zUtMGJeedEvhYcqjI5cBsgBJYpv6aYfKqJ0w2QDVanlpk8yERH8zheF2MLRGi2RFgS9RpkfCs5sBqgs4x47dAuGsbA0gnoyBWZ6cfBo9CS2y47RCwBD6PULjq-bcoVpONJsmybaRzg0dkHHZlUKU4ErGpGAECgODlwoChpw7FbIHXydUKnFiZzSOX4CQOAASYJIxU+CcViJKydXXDf5lHSJjOBxAYvHLcuEuVS7-qpRIDnNKUCqHKe6Ze9Nwf80NHtU5Xy-zJmfv08MAmExGPSAz8X4v0QMA4eLgsALyfjGCBEDkQfzIC4Nghx4A0mSL9Lgux66IKQZ1TkMUwGEKQa-FYaBEhpDJLOdundMQuHXnYFwoRjS1x-hQ0yUCzLUNoWguBoQEHcNEUwoReBZwPxwFwsRpl0LoWRGscwxxZGiM6uQ7hKDNroMwTSfhyI+AuFkC9AhciNEUIsUg3hNIlFH3BCIsR3s+ZcAFgHAiItxCkXFlRTRpkWCIEUFobRn9zh8TUZAYhihSHgK0U3GhxAIlWP1BojR3NULAneGiDEkJoSfD4OTSA2Iqg1CQKgXAFI7KFKUaCQoLwBQUHSRUykeATBax+BUfWbBDbXmNtOU25d7JW2YrbZMZYHZWhdsDN2H11z73EHtLpNJxyMCoIwJ8jB9DrKYGgRgk4nzjlDA+TwD5qyTgIeAIAA

**Comment by pleerock** [comment]
I'm able to share a codebase (relatively large with many complex types in) which contains a reproduction of this error.

**Comment by jevonearth** [comment]
What's the likelihood of this issue getting addressed in the upcoming 4.2.0 release? Is someone actively working on this issue?

**Comment by jevonearth** [comment]
It seems the issue for us is that type aliases are not lazily evaluated, and the compiler tries to expand all combinations until it can't.

To work around this, we introduced an empty interface in places, as it seems they are lazy. It's a bit of a kludge, and also a bit awkward because interfaces can't be used in unions.

Is there a better workaround/approach that we can take today, or a change in the compiler/language that might help us?

**Comment by akutruff** [comment]
I just hit this too when using 4.1's recursive conditional types.  (Super amazing feature by the way.)  It's hard to pair down to a simple example by the nature of this type of error.  The below code allows someone to declare a TypeScript type.  At the center of it all is ```Types``` which is a recursive conditional type.  I had to remove constraints on ```UnionType<Types>``` to ```UnionTypes<unkown>```  Further, there's other places in the code where I should be able to add constraints to ```Types``` but I had to slowly remove them and add downcasts to preven this error from showing up.  Sorry, I know this is a lot of code, and doesn't exactly repro given what I have below, but it is the level of complexity I'm dealing with.

```typescript
export interface Type<T> {
    kind: T,
}

export const str: Type<'string'> = { kind: 'string', };
export const num: Type<'number'> = { kind: 'number', };
export const bool: Type<'boolean'> = { kind: 'boolean', };
export const bigint: Type<'bigint'> = { kind: 'bigint', };
export const nul: Type<'null'> = { kind: 'null', };
export const undef: Type<'undefined'> = { kind: 'undefined', };

export interface LiteralType<TLiteralValue> extends Type<'literal'> {
    kind: 'literal',
    literal: TLiteralValue
}

export interface ArrayType<TElement> extends Type<'array'> {
    kind: 'array',
    elementType: TElement
}

export interface UnionType<TMembers> extends Type<'union'> {
    kind: 'union',
    memberTypes: Array<TMembers>
}

export interface ObjType<TShapeDefinition> extends Type<'object'> {
    kind: 'object',
    shapeDefinition: TShapeDefinition;
}

export const literal = <T>(stringLiteral: StringLiteral<T>): LiteralType<T> => ({
    kind: 'literal',
    literal: stringLiteral,
});

export const array = <T extends Types>(element: T): ArrayType<T> => ({
    kind: 'array',
    elementType: element,
});

export function obj<TShapeDefinition extends IShapeDefinition>(shapeDefinition: TShapeDefinition): ObjType<TShapeDefinition> {
    return {
        kind: 'object',
        shapeDefinition
    };
}

export type OptionalType<T> = UnionType<T | typeof undef>;

//TODO: UnionType should be able to be UnionType<Types> but there is a depth hueristic that blows up.  
//    That means functions and types have to cast until it's fixed
//     https://github.com/microsoft/TypeScript/issues/34933
export type Types =
    Type<'string'> |
    Type<'number'> |
    Type<'boolean'> |
    Type<'bigint'> |
    Type<'null'> |
    Type<'undefined'> |
    LiteralType<unknown> |
    ArrayType<Types> |
    UnionType<Types> |  //Should be able to be UnionType<Types>, but has to be UnionType<unknown> to avoid this error!  
    ObjType<IShapeDefinition>;

export interface IShapeDefinition {
    [index: string]: Types
}

export type Shape<TDefinition> =
    TDefinition extends LiteralType<infer LiteralKind>
    ? LiteralKind
    : TDefinition extends Type<'string'>
    ? string
    : TDefinition extends Type<'number'>
    ? number
    : TDefinition extends Type<'boolean'>
    ? boolean
    : TDefinition extends Type<'bigint'>
    ? bigint
    : TDefinition extends Type<'null'>
    ? null
    : TDefinition extends Type<'undefined'>
    ? undefined
    : TDefinition extends ArrayType<infer ElementKind>
    ? Array<Shape<ElementKind>>
    : TDefinition extends UnionType<infer KeyKind>
    ? Shape<KeyKind>
    : TDefinition extends ObjType<infer TShapeDefinition>
    ? { [P in keyof TShapeDefinition]: Shape<TShapeDefinition[P]> }
    : never;

type ShapeDefinitionDistribute<TsType> =
    TsType extends StringLiteral<TsType>
    ? LiteralType<TsType>
    : TsType extends string
    ? Type<'string'>
    : TsType extends number
    ? Type<'number'>
    : TsType extends boolean
    ? Type<'boolean'>
    : TsType extends bigint
    ? Type<'bigint'>
    : TsType extends null
    ? Type<'null'>
    : TsType extends undefined
    ? Type<'undefined'>
    : TsType extends Array<infer TElementType>
    ? ArrayType<ShapeDefinition<TElementType>>
    : TsType extends object
    ? ObjType<{ [P in keyof TsType]: ShapeDefinition<TsType[P]> }>
    : never;

export type ShapeDefinition<TsType> =
    NotAUnion<TsType> extends never
    ? UnionType<ShapeDefinitionDistribute<TsType>>
    : ShapeDefinitionDistribute<TsType>;

//mad science: https://stackoverflow.com/questions/50639496/is-there-a-way-to-prevent-union-types-in-typescript 
export type NotAUnion<T, U = T> = U extends any ? [T] extends [boolean] ? T : [T] extends [U] ? T : never : never;

export type StringLiteral<T> = T extends `${string & T}` ? T : never;

export type FlattenedUnion<T> =
    T extends UnionType<infer K>
    ? FlattenedUnion<K>
    : T;

export type CollapseSingleMemberUnion<T> =
    NotAUnion<T> extends never
    ? UnionType<T>
    : T;

export const flattenMembers = <T extends Types[]>(members: T): Array<FlattenedUnion<T>> => {
    return {} as any as Array<FlattenedUnion<T>>;  //code removed for this example
};

export const union = <T extends Types[]>(members: T): CollapseSingleMemberUnion<FlattenedUnion<T>> => {
    const flat = flattenMembers(members);  //Blows up when UnionType sometimes
    return {} as CollapseSingleMemberUnion<FlattenedUnion<T>>;  // edited for this example
};

const blowsUpSomeTypes = union([obj({name: str, age: num}), obj({foo: bool, age: num})]);
```

**Comment by andrei9669** [comment]
just hit this error when trying to make a generic interface for an api that uses only post request.
that api has a ton of request variations. I guess that did the thing :/
I hit the limit after only 17 requests(51 if I include all possible request param combinations).

**Comment by akutruff** [comment]
Found this comment in src/compiler/checker.ts
```typescript
   function getConditionalType(root: ConditionalRoot, mapper: TypeMapper | undefined, aliasSymbol?: Symbol, aliasTypeArguments?: readonly Type[]): Type {
            let result;
            let extraTypes: Type[] | undefined;
            // We loop here for an immediately nested conditional type in the false position, effectively treating
            // types of the form 'A extends B ? X : C extends D ? Y : E extends F ? Z : ...' as a single construct for
            // purposes of resolution. This means such types aren't subject to the instatiation depth limiter.
   /*....*/
}
```

I removed as many intermediate templates as possible, and it started to help.  Though, it's probably just kicking the can down the road and it greatly hurt readability as I have to put all my type mappings inline: 

```typescript

ObjType<{
        [P in keyof TsType]-?: Prop<
            ToDefinitionType<Exclude<TsType[P], undefined>>,
            P extends {
                [KK in keyof TsType]-?: undefined extends TsType[KK] ? KK : never
            }[keyof TsType] ? true : false, false>
    }>
```

**Comment by sandorvasas** [comment]
I hav two mongoose type definitions A and B. Both are identical, except field types (e.g. number in A, string in B, only primitive types differ). A compiles fine, B throws this 'Type instantiation is excessively deep and possibly infinite.' :DDDDDDDDDDDD

Hilarious compiler.

**Comment by DanielSWolf** [comment]
I'm using a parser combinator to combine generic parser functions into more complex parsers. As soon as my parser expressions exceed a certain length (which I don't find excessive), this error occurs, forcing me to re-design my parsers.

The frustrating aspect about this is that the error message gives no indication of *where* it sees excessive complexity, forcing me to use trial and error to fix the problem.

I'd be happy to supply repro code if it helps.

**Comment by luisgregson** [comment]
> just hit this error when trying to make a generic interface for an api that uses only post request.
> that api has a ton of request variations. I guess that did the thing :/
> I hit the limit after only 17 requests(51 if I include all possible request param combinations).

For this particular use case i wouldn't not using a generic interface but instead type the body as `unknown` and then using a type assertion for each variation. You could abstract out some of the shared assertion logic (and corresponding types) to make it less verbose.

Would this work for you or is there a reason that a Generic would be better?

**Comment by noam-honig** [comment]
Just hit this problem when using createSlice in @reduxjs/toolkit.

I was able to distill it to a stand-alone typescript file:
```
declare type Draft<T> = WritableDraft<T>;
declare type WritableDraft<T> = {
    -readonly [K in keyof T]: Draft<T[K]>;
};
declare type SliceCaseReducers<State> = {

    [K: string]: (state: Draft<State>) => State;
};

function x<CaseReducers extends SliceCaseReducers<{ test?: classWithRecursiveType[]; }>>(what: CaseReducers) {
}

x<{
    addComment: (state: WritableDraft<{
        test?: classWithRecursiveType[];
    }>) => {
        test: WritableDraft<classWithRecursiveType>[];
    };
}>({} as any);

declare type RecursiveType = string | RecursiveType[];

class classWithRecursiveType {
    b: string;
    a: RecursiveType; //comment me out to prevent the error
}
```

In this code, I'm getting an error for the line that starts with "x<{..." Type instantiation is excessively deep and possibly infinite.ts(2589)

If I comment out the line that starts with "a:..." the error stops.

**Comment by nawlbergs** [comment]
@noam-honig 

 `declare type RecursiveType = string | RecursiveType[];`
Can you do this kinda of stuff in ts?   ... that actually seems like a valid error to me.

That aside... our pipeline just suddenly started throwing this.  The code changes between when it worked and stopped working is extremely minimal and leads me to believe a package update has caused it.  Our usage of TS is light and (since we dont know TS that well) have not done a lot of crazy extending, generics or whatnots.

**Comment by noam-honig** [comment]
Yes you can, it was added sometime ago

**Comment by nawlbergs** [comment]
So to anyone who lands here and is getting this error in Node with express.

Our pipeline suddenly broke.  No major code changes on our part.. and it appears npm updated some packages
- Typescript went from `4.0.2` to `4.3.4` (i think)
- @types/node from `14.14.22` to `14.14.44`
- @types/express from `4.17.11` to `4.17.12`
- and a dependency of @types/express is @type/express-serve-static-core... which apparently when from `4.17.19` to `4.17.22`

Running `npx tsc` was after these updates was causing the `deep and infinite` error.  And very strangely..
for instance:
```js
const router = express.router();
router.get('/v3/app/userOpportunities/:userOpportunity_id/roadmaps', [...]); // <-- this was one of the error lines
export { router }
```
and even more odd.... is that this STILL throws the error
```js
...
router.get(':userOpportunity_id'); // still an error
...
```
and so I started playing around:
```js
...
router.get(':userOpportunity'); // error
router.get('userOpportunity'); // no error (removed the semi colon)
router.get(':userOpp'); // no error changed word
```

having `:userOpportunity` anywhere in my route was throwing the error. but `userOpportunity`, `:userOpp`, `something else` etc would be fine.

I had three routes throwing errors.. all of them were acting similar but with different words.. 2 routes had `:userOpportunity` it and the third was complaining about `:recommendationId`

This made me think that the express types were being naughty.  I started rolling things back slowly.  Thus, I added `@type/express-serve-static-core` to my package.json (it was just a dependency) and forced it back to `4.17.11` and now the TS errors are gone.

On a side note.. this only seems to fail in a unix environment for some reason.

**Comment by cmlenz** [comment]
For what it's worth, I also hit this error when using `@reduxjs/toolkit`, but specifically when using `immer` in reducers in `createSlice` et al.

**Comment by shanehsu** [comment]
@nawlbergs The new typings for express uses new TypeScript features in 4.0 to do wonders. See [this line from its typings](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a46c575be7a4f3d9ddc51b2e0938cf3855b68b4d/types/express-serve-static-core/index.d.ts#L101). 

It's like `constexpr` where we generate information from source, so we can get statically-typed `req.params` from the path that we write. Unfortunately, these magic requires TypeScript compiler to do more work, and the TypeScript compiler at its current form cannot quite cope with it. This is the root cause of the problem.

Still, +1 for both TypeScript and Express/Typings team to make the world more type-safe.

> Chimed in because I almost wasted with my own implementation, and also want to let more people know about this feature of TypeScript and how Express is not trying to be naughty, but more useful. 

Hope this gets fixed soon.

**Comment by nawlbergs** [comment]
@shanehsu 

I agree, +1 to the typescript team and the team behind DefinitelyTyped.  It's a huge boon to the JS community.  Not a fan of weird bugs that make no sense to me and break our entire app however :P

**Comment by robsco-git** [comment]
> This made me think that the express types were being naughty. I started rolling things back slowly. Thus, I added `@type/express-serve-static-core` to my package.json (it was just a dependency) and forced it back to `4.17.21` and now the TS errors are gone.

Thanks for the tip. I installed and setup https://github.com/rogeriochaves/npm-force-resolutions and added this to `package.json`:

```json
"resolutions": {
  "@types/express-serve-static-core": "4.17.21"
}
```
After a `npm install` the issue is resolved.

**Comment by jjangga0214** [comment]
# tsconfig

I hope to adjust the limit by tsconfig. How do you think?

**Comment by tavurth** [comment]
@nawlbergs thank you!

I was using `:orderReferenceNumber` in `tsoa` and it was breaking, but changing it to `:orderReference` fixed my issue :)

**Comment by jjangga0214** [comment]
## Guys, please show your interest in #44997 (thumb up!) to resolve the issue faster 

Currently, the limitation thresholds are hard-coded, to `50` and `5000000`.

***src/compiler/checker.ts***
```ts
if (
  instantiationDepth === 50 ||
  instantiationCount >= 5000000
) {
  // We have reached 50 recursive type instantiations and there is a very high likelyhood we're dealing
  // with a combination of infinite generic types that perpetually generate new type identities. We stop
  // the recursion here by yielding the error type.
  error(
    currentNode,
    Diagnostics.Type_instantiation_is_excessively_deep_and_possibly_infinite
  );
  return errorType;
}
```

To make this comfigurable, I pushed a PR #44997, which makes those values be able to be set by `compilerOptions`, which can be passed through CLI options or `tsconfig.json`, like an example below.

```json
{
  "compilerOptions": {
    "instantiationDepthLimit": 100,
    "instantiationCountLimit": 10000000
  }
}
```

**Comment by kiprasmel** [comment]
in the meantime, one can use [patch-package](http://npm.im/patch-package) to create a per-project patch to implement [the temporary fix](https://github.com/microsoft/TypeScript/issues/34933#issuecomment-552500444):

1. edit `node_modules/typescript/lib/tsc.js`: search for `5000000` or `instantiationCount` and increate both the `instantiationDepth` and `instantiationCount` limits

2. run `node_modules/.bin/patch-package typescript` (via `yarn patch-package typescript` or `npx patch-package typescript` to generate the patch

should end up something like this in the `patches/` folder:

```diff
diff --git a/node_modules/typescript/lib/tsc.js b/node_modules/typescript/lib/tsc.js
index 78c5ec2..7d6c586 100644
--- a/node_modules/typescript/lib/tsc.js
+++ b/node_modules/typescript/lib/tsc.js
@@ -49264,7 +49265,7 @@ var ts;
             if (!couldContainTypeVariables(type)) {
                 return type;
             }
-            if (instantiationDepth === 50 || instantiationCount >= 5000000) {
+            if (instantiationDepth === (50 * 2) || instantiationCount >= (5000000 * 2)) {
                 ts.tracing === null || ts.tracing === void 0 ? void 0 : ts.tracing.instant("checkTypes", "instantiateType_DepthLimit", { typeId: type.id, instantiationDepth: instantiationDepth, instantiationCount: instantiationCount });
                 error(currentNode, ts.Diagnostics.Type_instantiation_is_excessively_deep_and_possibly_infinite);
                 return errorType;

```

3. make sure to always apply the patch (e.g. after installation) by following [patch-package's docs for this purpose](https://github.com/ds300/patch-package#set-up) (also note [this](https://github.com/ds300/patch-package#why-use-postinstall-postinstall-with-yarn)).

😈

**Comment by dstoyanoff** [comment]
Have the same issue. The interesting part is that I have 2 variables with the same type and one of them works, while the other throws this error. 

Upgrading to 4.4.2 fixes the issue for me.

**Comment by Ricardo-Marques** [comment]
The depth fix mentioned in this article resolved my problem, even when sticking to an initial depth of 1
https://www.angularfix.com/2022/01/why-am-i-getting-instantiation-is.html

**Comment by ramblehead** [comment]
It is possible to avoid Subj. issue in certain use cases by utilising [UnionToIntersection](https://stackoverflow.com/a/50375286/5000057) metafunction as shown in the following SO answer:

https://stackoverflow.com/a/74292294/5000057

**Comment by thelinuxlich** [comment]
What about raising the instantiationDepth limit to 500?

**Comment by iway1** [comment]
50 is really not that deep if you're using fancy typescript inference... upping the limit could be pretty nice

**Comment by MohammadKurjieh** [comment]
Any updates on this issue?

**Comment by Xriuk** [comment]
This is definitely a problem for us, as we make an intensive use of type inference, and the type sometimes works and sometimes it produces this error, it's not recursive, maybe simply too complex/deep.
Also an indication about where (in the nested types) this error happens would be useful, as we cannot determine if it is really an error on our side or just TypeScript limitations

**Comment by rossPatton** [comment]
Hitting this issue. Using zod to merge several sub-schemas into a "complete" schema, resulting in the error

**Comment by OnkelTem** [comment]
Was also experimenting with Zod schemas. In my case, this would negate all my efforts to validate the API using Zod.

**Comment by andelkocvjetkovic** [comment]
I had similar issues with zod, instead using z.merge I used spread operator
```ts
export const addNewProspectSchema = z.object({
  ...generalSchema.shape,
  ...landSchema.shape,
  ...pvResourceSchema.shape,
  ...windResourceSchema.shape,
  ...productionSchema.shape,
  ...gridSchema.shape,
  ...strategySchema.shape,
  ...scheduleSchema.shape,
  ...documentsAndRedFlagsSchema.shape,
  ...bussinessDevBudgetAndStepsSchema.shape,
  ...financialPerformanceSchema.shape,
  ...environmentSchema.shape,
  ...developmentMilestonesSchema.shape,
});

type AddNewProspectSchema = z.infer<typeof addNewProspectSchema>;

```

This solved the problem at least for now. :)

**Comment by ryan-williams** [comment]
In case it helps someone, I think I was triggering this error attempting to combine some `Promise<JSX.Element>`s with `Promise<ReactNode>`s in one array:

```ts
const elems: Promise<JSX.Element>[] = []
const nodes: Promise<ReactNode>[] = []
const all = Promise.all([ ...elems, ...nodes ])
```

This fails to compile:
```
Type error: Type instantiation is excessively deep and possibly infinite.

  112 |     const elems: Promise<JSX.Element>[] = []
  113 |     const nodes: Promise<ReactNode>[] = []
> 114 |     const all = Promise.all([ ...elems, ...nodes ])
      |                 ^
```

In my case, both arrays could be refined to `Promise<JSX.Element>`s, and that fixed the error. but in general the other direction ("widening" `JSX.Element`s to `ReactNode`s) should always work.

**Comment by Amorim33** [comment]
> I had similar issue with zod, instead using z.merge or z.extend I used spread operators
> 
> ```ts
> export const addNewProspectSchema = z.object({
>   ...generalSchema.shape,
>   ...landSchema.shape,
>   ...pvResourceSchema.shape,
>   ...windResourceSchema.shape,
>   ...productionSchema.shape,
>   ...gridSchema.shape,
>   ...
> });
> 
> type AddNewProspectSchema = z.infer<typeof addNewProspectSchema>;
> ```
> 
> This solved the issue, at least for now. :)

Very nice!!

I have a similar problem.

I was using merge, pick and extend. After seeing your comment, I simplified the schema using the spread.
However, the error continues to appear, probably the cause is now the pick method.

Any ideas on how to resolve it?

```ts
const getMonthlyInsuredItemSchema = monthlyInsuredItemSchema.pick({
  month: true,
  processedAt: true,
});

const getInsuredItemSchema = insuredItemSchema.pick({
  externalId: true,
  additionalAttributes: true,
  remainingInstallments: true,
  totalInstallments: true,
  installmentAmount: true,
});

const getInsuredPersonSchema = insuredPersonSchema.pick({
  name: true,
  cpf: true,
  birthdate: true,
  gender: true,
});

const getInsuredItemsResponse = z.array(
  z.object({
    ...getMonthlyInsuredItemSchema.shape,
    ...getInsuredItemSchema.shape,
    insuredPerson: getInsuredPersonSchema,
  }),
);
```

**Comment by natew** [comment]
One point in favor of upping the limit besides the weeks of my life wasted that I've spent battling these problems, is that I actually could easily speed up the [tamagui](https://github.com/tamagui/tamagui) types by 2x, except that it starts hitting complexity limits.

So in fact the complexity limit is making things slower in practice rather than faster.

**Comment by aleczratiu** [comment]
> Hitting this issue. Using zod to merge several sub-schemas into a "complete" schema, resulting in the error. Hard-coding the resultant large schema works fine, but is more of a pain to maintain/less re-usable

The same issue I have using zod, did you find any workarounds for it?

**Comment by Charismara** [comment]
> > Hitting this issue. Using zod to merge several sub-schemas into a "complete" schema, resulting in the error. Hard-coding the resultant large schema works fine, but is more of a pain to maintain/less re-usable
> 
> The same issue I have using zod, did you find any workarounds for it?

I got the same issue. Is there a workaround for this?

**Comment by Threebow** [comment]
What I interpret from this issue, and the comments within, is that that once your codebase/typing logic grows enough in size or complexity to an arbitrary point that TypeScript is unhappy with, you are no longer able to compile the project.

So, am I correct to conclude from this, that TypeScript enforces an upper limit on the size of project that you are able to develop with it, and any larger/more complex projects are entirely unsupported by the compiler?

Somebody please correct me if I am wrong, but the fact that you can spend years working on a project and one day hit an arbitrary "complexity limit" seems grossly incorrect and inappropriate for a tool of this breadth in the industry.

If this is indeed the case, and I am not mistaken, then:

1. What is the accepted "fix" to spending years on a project just to ultimately hit this limit and be faced with a compiler that outright refuses to compile the project? Are we doomed to throw the project out and rewrite the whole thing, or put an indefinite pause on development until this issue is fixed?
2. Where does TypeScript make this clear enough to a potential user looking to adopt TypeScript, to where they will choose a different tool for a project if they know their project will grow to this arbitrary complexity limit, to completely avoid wasting years of their and their teammates' time?

Open to input and corrections from others who have bumped into this.

**Comment by natew** [comment]
I think this really needs an option to just increase the size.

The reason this is such a big problem is that - only the biggest projects are affected.

IE, you've spent the *most* amount of effort investing into TypeScript, to the point where your types are really large. But now it actually comes around and bites you, and often in a way you just can't control. If a library you use everything in your project has complex types and you've used it 2k times across your codebase, and on the 2001 time it suddenly craps out, you're looking at losing type safety on one of your most critical/used pieces of the stack.

The second reason I think we need a way around this is because the warning actually doesn't correlate with performance. I've landed big improvements to performance that caused this warning to occur. So it actually is a safeguard that is working against you oftentimes.

**Comment by Threebow** [comment]
@natew 

> I think this really needs an option to just increase the size.

[TS has stated five years ago in 2019 that they will not expose this option to users because it's an implementation detail of an internal TypeScript mechanism that they may choose to eliminate in the future.](https://github.com/microsoft/TypeScript/issues/29511#issuecomment-458620791)

[Another person has said the same thing, over a year later, in 2020.](https://github.com/microsoft/TypeScript/pull/29602#issuecomment-581545268)

These decisions seem a little short-sighted in hindsight, given that it's been five years, and that they have seemingly forgotten about this mechanism in its entirety.

> IE, you've spent the most amount of effort investing into TypeScript, to the point where your types are really large. But now it actually comes around and bites you, and often in a way you just can't control.

I strongly resonate with this. It feels like a kick in the groin to hear that your usage of the system is outside of what they support, after investing so heavily in the ecosystem and the language itself.

> If a library you use everything in your project has complex types and you've used it 2k times across your codebase, and on the 2001 time it suddenly craps out, you're looking at losing type safety on one of your most critical/used pieces of the stack.

This is simply unacceptable. This issue is only going to start popping up more and more, as more people start to get into codegen of complicated type systems.

**Comment by tran-simon** [comment]
I get this issue all the time beacuse I'm using [json-schema-to-ts](https://www.npmjs.com/package/json-schema-to-ts)

[It's documented in their FAQ](https://github.com/ThomasAribart/json-schema-to-ts/blob/main/documentation/FAQs/i-get-a-type-instantiation-is-excessively-deep-and-potentially-infinite-error-what-should-i-do.md), but there doesn't seem to be any fix

**Comment by Xriuk** [comment]
> What I interpret from this issue, and the comments within, is that that once your codebase/typing logic grows enough in size or complexity to an arbitrary point that TypeScript is unhappy with, you are no longer able to compile the project.

More or less, the main problem with increasing type instantiation depth appears to be that some web instances of TypeScript just crash, because TypeScript is meant to be running everywhere apparently.
This is frustrating because at least for me TypeScript just compiles code into JavaScript, so no matter the complexity and depth of types, the resulting code should always be the same. That's why we resorted to using a custom TypeScript version with [these values](https://github.com/microsoft/TypeScript/issues/34933#issuecomment-552500444) increased freely.

**Comment by Threebow** [comment]
> some web instances of TypeScript just crash, because TypeScript is meant to be running everywhere apparently

Now I am curious if there is a limit to where TypeScript crashes simply because the compilation time exceeds an arbitrary "too long to compile" point. It seems strange that a long compilation time would force a crash, as opposed to a simple freeze.

That seems like a separate issue in its entirety, unrelated to this one. I'm also not really sure who's compiling a project with 2000+ Zod usages in their browser at runtime...

**Comment by vipcxj** [comment]
same issue，happened when using redux-tookit. what a horrible bug. any news?

**Comment by ChapelStudios** [comment]
Still nothing? the milestone is listed as 5.7 but TS is on 5.9 now. I'm running into this on a smaller project with a lot of class inheritance and class mix-ins.
